libp2p/connmanager.nim

```diff
@@ -0,0 +1,276 @@
+
+  return conn
+
+proc selectMuxer*(c: ConnManager, conn: Connection): Muxer =
+  ## select the muxer for the provided connection
+  ##
+
+  if isNil(conn):
+    return
+
+  if conn in c.muxed:
+    return c.muxed[conn].muxer
+
+proc storeConn*(c: ConnManager, conn: Connection) =
+  ## store a connection
+  ##
+
+  if isNil(conn):
+    raise newException(CatchableError, "connection cannot be nil")
+
+  if isNil(conn.peerInfo):
+    raise newException(CatchableError, "empty peer info")
+
+  let peerInfo = conn.peerInfo
+  if c.conns.getOrDefault(peerInfo.peerId).len > c.maxConns:
+    trace "too many connections", peer = $conn.peerInfo,
+                                  conns = c.conns
+                                  .getOrDefault(peerInfo.peerId).len
+
+    raise newTooManyConnections()
+
+  if peerInfo.peerId notin c.conns:
+    c.conns[peerInfo.peerId] = initHashSet[Connection]()
+
+  c.conns[peerInfo.peerId].incl(conn)
+
+  # launch on close listener
+  asyncCheck c.onClose(conn)
+  libp2p_peers.set(c.conns.len.int64)
+
+proc storeOutgoing*(c: ConnManager, conn: Connection) =
+  conn.dir = Direction.Out
+  c.storeConn(conn)
+
+proc storeIncoming*(c: ConnManager, conn: Connection) =
+  conn.dir = Direction.In
+  c.storeConn(conn)
+
+proc storeMuxer*(c: ConnManager,
+                 muxer: Muxer,
+                 handle: Future[void] = nil) =
+  ## store the connection and muxer
+  ##
+
+  if isNil(muxer):
+    raise newException(CatchableError, "muxer cannot be nil")
+
+  if isNil(muxer.connection):
+    raise newException(CatchableError, "muxer's connection cannot be nil")
+
+  c.muxed[muxer.connection] = MuxerHolder(
+    muxer: muxer,
+    handle: handle)
+
+  trace "storred connection", connections = c.conns.len
+
+proc getMuxedStream*(c: ConnManager,
+                     peerInfo: PeerInfo,
+                     dir: Direction): Future[Connection] {.async, gcsafe.} =
+  ## get a muxed stream for the provided peer
+  ## with the given direction
+  ##
+
+  let muxer = c.selectMuxer(c.selectConn(peerInfo, dir)) # always get the first muxer here
+  if not(isNil(muxer)):
+    return await muxer.newStream()
+
+proc getMuxedStream*(c: ConnManager,
+                     peerInfo: PeerInfo): Future[Connection] {.async, gcsafe.} =
+  ## get a muxed stream for the passed peer from any connection
+  ##
+
+  let muxer = c.selectMuxer(c.selectConn(peerInfo)) # always get the first muxer here
+  if not(isNil(muxer)):
+    return await muxer.newStream()
+
+proc getMuxedStream*(c: ConnManager,
+                     conn: Connection): Future[Connection] {.async, gcsafe.} =
+  ## get a muxed stream for the passed connection
+  ##
+
+  let muxer = c.selectMuxer(conn) # always get the first muxer here
+  if not(isNil(muxer)):
+    return await muxer.newStream()
+
+proc dropPeer*(c: ConnManager, peerInfo: PeerInfo) {.async.} =
+  ## drop connections and cleanup resources for peer
+  ##
+
+  for conn in c.conns.getOrDefault(peerInfo.peerId):
```

Comment:
awaiting inside a iteration is a bit dangerous for race conditions , hmmm maybe `conns` should be interlocked too cos of this.

Reply:
Also `getOrDefault` is really a bad habit I started to realize.. I'd rather defect and understand why there was no record . doing that a lot in gossip 1.1

Reply:
> awaiting inside a iteration is a bit dangerous for race conditions , hmmm maybe `conns` should be interlocked too cos of this.

Not sure, maybe we want do lock `dropPeer`, we can also do an `allFinished` as well.